### ISSUE TITLE
Permisions for non-root users are lost on DinD containers when system container is restarted

### ISSUE THREAD

**Issue by joanbm** [nestybox/sysbox#251]
# Steps to reproduce

1. Set up a Ubuntu 20.04 + Docker install (`docker.io` package is enough, without userns-remap).

2. Install and start Sysbox 0.3.0 (so it is working in "Auto userns ID" mode) using the official .deb package.

3. Launch a Docker-in-Docker scenario with `/var/lib/docker` bound to a host path, here `/var/lib/sub-docker`:

```
host$ docker run --runtime=sysbox-runc -v /var/lib/sub-docker:/var/lib/docker --rm -it nestybox/alpine-docker
container# dockerd &
[...]
```

4. Launch a container which runs as a non-root user internally and has persistent storage on a volume, here I'm using `crunchydata/crunchy-postgres-appdev:centos7-12.3-4.3.2`. It will launch correctly:

```
container# docker volume create pgdata
container# docker run -it --rm -e PG_PASSWORD=password --hostname stuff -v pgdata:/pgdata crunchydata/crunchy-postgres-appdev:centos7-12.3-4.3.2
[...]
2021-04-06 17:33:15 UTC [137]: [2-1] user=,db=,app=,client=LOG:  listening on IPv4 address "0.0.0.0", port 5432
2021-04-06 17:33:15 UTC [137]: [3-1] user=,db=,app=,client=LOG:  listening on IPv6 address "::", port 5432
2021-04-06 17:33:15 UTC [137]: [4-1] user=,db=,app=,client=LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
```

(The Postgres container can also be stopped and restarted successfully)

6. Stop both the Postgres and the Docker-in-Docker (nestybox/alpine-docker) container.

7. Restart the Docker-in-Docker system container and Postgres containers. Now it will unexpectedly to start:

```
host$ docker run --runtime=sysbox-runc -v /var/lib/sub-docker:/var/lib/docker --rm -it nestybox/alpine-docker
container# dockerd &
[...]
container# docker run -it --rm -e PG_PASSWORD=password --hostname stuff -v pgdata:/pgdata crunchydata/crunchy-postgres-appdev:centos7-12.3-4.3.2
Tue Apr  6 17:34:08 UTC 2021 ERROR: Initializing the database (initdb): Unable to initialize the database: 
initdb: error: could not access directory "/pgdata/stuff": Permission denied
```

# Explanation

The problem appears to happen on the following circumstances:

1. Using "Auto userns ID" mode
2. Using Docker-in-Docker by binding /var/lib/docker
2. The container running inside DinD runs as a non-root user which owns a volume
4. The container does not have any logic to "re-own" the volume (it appears many popular containers do)

While the container is running, `/var/lib/sub-docker/` is owned by sub-container root = DinD uid 100000 = 'sysbox' subuid/subgid  and `/var/lib/sub-docker/volumes/pgdata/_data` is owned by sub-container postgres = 100000  + 26:
```
root@ubuntu-s-1vcpu-2gb-fra1-01:~# ls -lah /var/lib/sub-docker/
total 60K
drwx--x--x 15 100000 100000 4.0K Apr  6 17:31 .
drwxr-xr-x 44 root   root   4.0K Apr  6 17:30 ..
drwx------  2 100000 100000 4.0K Apr  6 17:31 builder
drwx--x--x  4 100000 100000 4.0K Apr  6 17:31 buildkit
drwx--x--x  3 100000 100000 4.0K Apr  6 17:31 containerd
drwx------  3 100000 100000 4.0K Apr  6 17:33 containers
drwx------  3 100000 100000 4.0K Apr  6 17:31 image
drwxr-x---  3 100000 100000 4.0K Apr  6 17:31 network
drwx------ 27 100000 100000 4.0K Apr  6 17:33 overlay2
drwx------  4 100000 100000 4.0K Apr  6 17:31 plugins
drwx------  2 100000 100000 4.0K Apr  6 17:31 runtimes
drwx------  2 100000 100000 4.0K Apr  6 17:31 swarm
drwx------  2 100000 100000 4.0K Apr  6 17:33 tmp
drwx------  2 100000 100000 4.0K Apr  6 17:31 trust
drwx------  4 100000 100000 4.0K Apr  6 17:33 volumes
root@ubuntu-s-1vcpu-2gb-fra1-01:~# ls -lah /var/lib/sub-docker/volumes/pgdata/_data
total 12K
drwxrwxr-x  3 100026 100026 4.0K Apr  6 17:33 .
drwxr-xr-x  3 100000 100000 4.0K Apr  6 17:31 ..
drwx------ 20 100026 100026 4.0K Apr  6 17:33 stuff
```

When the system container is stopped all files will be now owned by root, which means the permissions have been destroyed:
```
root@ubuntu-s-1vcpu-2gb-fra1-01:~# ls -lah /var/lib/sub-docker/
total 60K
drwx--x--x 15 root root 4.0K Apr  6 17:33 .
drwxr-xr-x 44 root root 4.0K Apr  6 17:30 ..
drwx------  2 root root 4.0K Apr  6 17:31 builder
drwx--x--x  4 root root 4.0K Apr  6 17:31 buildkit
drwx--x--x  3 root root 4.0K Apr  6 17:31 containerd
drwx------  2 root root 4.0K Apr  6 17:34 containers
drwx------  3 root root 4.0K Apr  6 17:31 image
drwxr-x---  3 root root 4.0K Apr  6 17:31 network
drwx------ 25 root root 4.0K Apr  6 17:34 overlay2
drwx------  4 root root 4.0K Apr  6 17:31 plugins
drwx------  2 root root 4.0K Apr  6 17:33 runtimes
drwx------  2 root root 4.0K Apr  6 17:31 swarm
drwx------  2 root root 4.0K Apr  6 17:33 tmp
drwx------  2 root root 4.0K Apr  6 17:31 trust
drwx------  3 root root 4.0K Apr  6 17:34 volumes
root@ubuntu-s-1vcpu-2gb-fra1-01:~# ls -lah /var/lib/sub-docker/volumes/pgdata/_data
total 12K
drwxrwxr-x  3 root root 4.0K Apr  6 17:33 .
drwxr-xr-x  3 root root 4.0K Apr  6 17:31 ..
drwx------ 20 root root 4.0K Apr  6 17:33 stuff
```

Re-launching the container will try to restore the permissions but the files which were owned by UID 100026 are now owned by UID 100000:

```
root@ubuntu-s-1vcpu-2gb-fra1-01:~# ls -lah /var/lib/sub-docker/
total 60K
drwx--x--x 15 100000 100000 4.0K Apr  6 17:33 .
drwxr-xr-x 44 root   root   4.0K Apr  6 17:30 ..
drwx------  2 100000 100000 4.0K Apr  6 17:31 builder
drwx--x--x  4 100000 100000 4.0K Apr  6 17:31 buildkit
drwx--x--x  3 100000 100000 4.0K Apr  6 17:31 containerd
drwx------  2 100000 100000 4.0K Apr  6 17:34 containers
drwx------  3 100000 100000 4.0K Apr  6 17:31 image
drwxr-x---  3 100000 100000 4.0K Apr  6 17:31 network
drwx------ 25 100000 100000 4.0K Apr  6 17:34 overlay2
drwx------  4 100000 100000 4.0K Apr  6 17:31 plugins
drwx------  2 100000 100000 4.0K Apr  6 17:33 runtimes
drwx------  2 100000 100000 4.0K Apr  6 17:31 swarm
drwx------  2 100000 100000 4.0K Apr  6 17:33 tmp
drwx------  2 100000 100000 4.0K Apr  6 17:31 trust
drwx------  3 100000 100000 4.0K Apr  6 17:34 volumes
root@ubuntu-s-1vcpu-2gb-fra1-01:~# ls -lah /var/lib/sub-docker/volumes/pgdata/_data
total 12K
drwxrwxr-x  3 100000 100000 4.0K Apr  6 17:33 .
drwxr-xr-x  3 100000 100000 4.0K Apr  6 17:31 ..
drwx------ 20 100000 100000 4.0K Apr  6 17:33 stuff
```

In `/var/log/sysbox-mgr.log` those permission changes are logged:

```
time="2021-04-06 17:34:36" level=info msg="chowning /var/lib/sub-docker to 100000:100000 for 579e2c6a6029"
time="2021-04-06 17:34:36" level=info msg="done chowning /var/lib/sub-docker for 579e2c6a6029"
time="2021-04-06 17:34:58" level=info msg="reverting chown on /var/lib/sub-docker to 0:0 for 579e2c6a6029"
time="2021-04-06 17:34:59" level=info msg="done reverting chown on /var/lib/sub-docker for 579e2c6a6029"
```

It appears that when stopping the container, sysbox will chown all files to root, even if they are not owned by the container root, possibly what it should do is shift the owners by whatever the 'sysbox' subuid/subgid id (i.e. by 100000 in this scenario).

# Workaround

As a workaround, creating the directory and making it owned by the sysbox uid/gid before starting the system container, will avoid the problem since this prevents the recursive chown from running:

`sudo mkdir /var/lib/sub-docker && sudo chown 100000:100000 /var/lib/sub-docker`

**Comment by ctalledo** [comment]
Hi @joanbm, thanks for discovering the issue and for the detailed report and debug on it.

I agree with your problem description: for host directories that are mounted inside the container over what Sysbox considers "special directories" (e.g., "/var/lib/docker"), sysbox performs a temporary chown of the host volume contents such that they will show up a "root:root" inside the container. When the container is destroyed, Sysbox reverts that chown to whatever it was before.

The bug  is that I was incorrectly assuming that the contents of such dirs should always chown to "root:root" because they were being mounted over "/var/lib/docker", but this is wrong because as you demonstrated not everything under "/var/lib/docker" has "root:root" ownership (particularly stuff in the container's `/var/lib/docker/volumes/...` may not).

The solution is to perform the chown in a more clever way, by "shifting" the uid:gids via `chown` as you mentioned (since we can't unfortunately mount shiftfs on that directory as Docker won't work well if we do).

I have to take a look at how to do this efficiently, as this can directly impact container startup & stop time, which is critical that we keep low.

Sysbox uses `rsync` for the transfer, and on a quick look it appears there may be a way to ask rsync to shift the file ownership (apparently via the `--usermap` and `--groupmap` options, though I've not yet tried them).

**Comment by ctalledo** [comment]
FYI, the chown is done in sysbox-mgr, here:

When container is started:
https://github.com/nestybox/sysbox-mgr/blob/2f343c4728e53d3ebeabbe15c9953d422ddf0168/mgr.go#L684

When container is stopped:
https://github.com/nestybox/sysbox-mgr/blob/2f343c4728e53d3ebeabbe15c9953d422ddf0168/mgr.go#L344

In the past we did the chown using golang's filepath.Walk(), but it was a bit slow, so we decided to use the `chown(2)` command. However this command won't do what we need it to do now. Seems like we may need to do back to filepath.Walk() in order to implement the chown-based uid:gid shifting.

**Comment by ctalledo** [comment]
Here is a reference example of how the good folks at LXD do this:

https://github.com/lxc/lxd/blob/0d62f01872c2fc3946e4b74bdc4c981654ee80de/shared/idmap/idmapset_linux.go#L512

**Comment by ctalledo** [comment]
I believe this issue must also be affecting Sysbox in scenarios where we inner container images are preloaded into a system container, as a similar chown operation is done (though it's a copy + chown done via rsync). Thus, the solution should also look into fixing not only the scenario originally reported, but this one too.

**Comment by ctalledo** [comment]
FYI, this PR has the fix: https://github.com/nestybox/sysbox-mgr/pull/26

**Comment by ctalledo** [comment]
PR merged, the issue is now fixed.

Thanks @joanbm for reporting this. Please give it a try and let me know of any more problems.

I'll close this issue, please re-open if needed.

**Comment by joanbm** [comment]
Hello,

Thank you very much for the quick fix! However, I was trying to verify the fix using the steps in the first post (except building from master, of course) and it doesn't appear to be working correctly for that case:

After stopping both the PostgreSQL container and the DinD system container, it fails to chown one of the files:

```
time="2021-04-17 18:57:17" level=info msg="registered new container 59a31fdff15e"
time="2021-04-17 18:57:17" level=info msg="shifting uids at /var/lib/sub-docker for 59a31fdff15e"
time="2021-04-17 18:57:17" level=info msg="done shifting uids at /var/lib/sub-docker for 59a31fdff15e"
time="2021-04-17 18:59:04" level=info msg="reverting uid-shift on /var/lib/sub-docker for 59a31fdff15e"
time="2021-04-17 18:59:04" level=warning msg="failed to revert uid-shift of mount source at /var/lib/sub-docker: failed to shift ACL for /var/lib/sub-docker/overlay2/13bb1753407bcc9d9aa857e8a4753388b6215edc5e0e9a14f10cae5e92aee969/diff/usr/lib64/libopcodes-2.27-43.base.el7.so: open /var/lib/sub-docker/overlay2/13bb1753407bcc9d9aa857e8a4753388b6215edc5e0e9a14f10cae5e92aee969/diff/usr/lib64/libopcodes-2.27-43.base.el7.so: no such device or address"
time="2021-04-17 18:59:04" level=info msg="done reverting uid-shift on /var/lib/sub-docker for 59a31fdff15e"
time="2021-04-17 18:59:04" level=info msg="unregistered container 59a31fdff15e"
time="2021-04-17 18:59:04" level=info msg="released resources for container 59a31fdff15e"
```

This file appears to have some funky properties so it's probably some edge case that is not working well with the chown algorithm:
```
root@ubuntu-s-2vcpu-2gb-intel-fra1-01:~# sudo ls -lah /var/lib/sub-docker/overlay2/13bb1753407bcc9d9aa857e8a4753388b6215edc5e0e9a14f10cae5e92aee969/diff/usr/lib64/libopcodes-2.27-43.base.el7.so
c--------- 1 100000 100000 0, 0 Apr 17 18:58 /var/lib/sub-docker/overlay2/13bb1753407bcc9d9aa857e8a4753388b6215edc5e0e9a14f10cae5e92aee969/diff/usr/lib64/libopcodes-2.27-43.base.el7.so
```

The folder has a mixture of root and container-root owned files since the chown process stops:
```
root@ubuntu-s-2vcpu-2gb-intel-fra1-01:~# ls -lah /var/lib/sub-docker/
total 60K
drwx--x--x 15 root   root   4.0K Apr 17 18:57 .
drwxr-xr-x 44 root   root   4.0K Apr 17 18:57 ..
drwx------  2 100000 100000 4.0K Apr 17 18:57 builder
drwx--x--x  4 100000 100000 4.0K Apr 17 18:57 buildkit
drwx--x--x  3 100000 100000 4.0K Apr 17 18:57 containerd
drwx------  2 100000 100000 4.0K Apr 17 18:58 containers
drwx------  3 root   root   4.0K Apr 17 18:57 image
drwxr-x---  3 100000 100000 4.0K Apr 17 18:57 network
drwx------ 25 root   root   4.0K Apr 17 18:58 overlay2
drwx------  4 100000 100000 4.0K Apr 17 18:57 plugins
drwx------  2 100000 100000 4.0K Apr 17 18:57 runtimes
drwx------  2 100000 100000 4.0K Apr 17 18:57 swarm
drwx------  2 100000 100000 4.0K Apr 17 18:58 tmp
drwx------  2 root   root   4.0K Apr 17 18:57 trust
drwx------  3 100000 100000 4.0K Apr 17 18:58 volumes
```

(As far as I can see I can't re-open the issue myself)

**Comment by joanbm** [comment]
It appears that the problem is with the `os.Open` call in the PR, calling it on that character device that is created inside the DinD environment fails. It appears to work by removing`os.Open` call and using the path-based instead of file-based operations in go-acl. It's probably a bit less optimal, though.

**Comment by ctalledo** [comment]
Thanks Joan; let's re-open to investigate and fix this corner case then.

**Comment by ctalledo** [comment]
This PR addresses this last issue reported by @joanbm:

https://github.com/nestybox/sysbox-mgr/pull/28

**Comment by ctalledo** [comment]
Hi @joanbm, I've fixed the latest issue you reported (as well as a performance regression caused by the original fix to this issue).

Let me know if it looks good please. Thanks!

**Comment by joanbm** [comment]
Hi @ctalledo, thank you very much! I've checked the issue with the updated patches from PRs #27 and #28 and it's working correctly now, so I'm closing the issue.

**Comment by ctalledo** [comment]
> Hi @ctalledo, thank you very much! I've checked the issue with the updated patches from PRs #27 and #28 and it's working correctly now, so I'm closing the issue.

Great! Thanks again for reporting it.
